### PR TITLE
Rework code review section

### LINKED ISSUE #217: Make the code review section more engaging
Currently the code review section ends with a long bullet point list of best practices. I propose restructuring this section:

* Go through technical steps to raise a review
* Get user to raise a review based on an example branch that will be included in the repo template
* Go through technical steps of adding a comment, get user to add a comment
* Cover some of the things that a review should look for
* Get user to go through the code looking for those things
* Repeat last two points if there are more than 6 things to look for
* Suggest a different person then attempts to address their review comments (or address their own if doing on own)
* Talk about empathy (and any other review considerations)

Steps to do this:
 
 - [x] Identify the kind of things to look for in a review
 - [x]  Propose new structure for the section
 - [x] Develop a piece of code that exhibits the problems
 - [x] Write the exercises and example solutions
 - [ ] Add new conent
 - [ ] Modify the content for creating a new pull request (this is quite different)
 - [ ] Modify existing content
 - [ ] Read through and make final tweaks
(I intend to address this, raising issue for visibility and in case anyone disagrees).

### PR DESCRIPTION
**Pull request by thomaskileyukaea**
Changes to the code review section, fixes #217 

The exercises are going to have the reader be reviewing a pull request identical to this one: https://github.com/thomaskileyukaea/python-intermediate-inflammation/pull/3

One major outstanding issue is how to get the example branch onto the users repo (templates don't copy branches in a way that would work). 

It may be easiest to see the changes commit by commit, as I did the structural changes first, before adding new things. 

- [x]  Rewrite ` Code Review and Pull Requests In Action` for the new exercise
- [x]  Update branch diagram for new branch situation (or remove)
- [x] Update screen shot for the new pull request
- [x] Update `Raising a pull request` for the new exercise
- [x] Screenshot todos
- [x] Fix numbering
- [x] Tidy up or remove `Best Practice for Code Review` 
- [x] Line by line edit of new content
- [x] SemBr pass
- [x] Final read through

### CODE REVIEW

File: _episodes/41-code-review.md [hunk 1]
```diff
@@ -65,32 +65,6 @@ and something you should adopt in your development process too.
...
-> > Developers use code review feedback from more senior developers
-> > to improve their own coding practices and expertise.
-> > Finally, it helps increase the sense of collective code ownership and responsibility,
-> > which in turn helps increase the "bus factor"
-> > and reduce the risk resulting from information and capabilities
-> > being held by a single person "responsible" for a certain part of the codebase
-> > and not being shared among team members.
-> {: .solution}
-{: .challenge}
-
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:93]: I didn't think this exercise added much and I wanted to add other more exercises - but I'm happy to discuss.

File: _episodes/41-code-review.md [hunk 2]
```diff
@@ -109,49 +83,6 @@ Since the cost of bug fixes grows in orders of magnitude throughout the software
...
-  which typically helps with the following tasks:
-  (1) collecting and displaying the updated files and highlighting what has changed,
-  (2) facilitating a conversation between team members (reviewers and developers), and
-  (3) allowing code administrators and product managers
-  a certain control and overview of the code development workflow.
-  Modern tools may provide a handful of other functionalities too, such as metrics
-  (e.g. inspection rate, defect rate, defect density).
-
-Each of the above techniques have their pros and cons and varying degrees practicality -
-it is up to the team to decide which ones are most suitable for the project and when to use them.
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:154]: I don't see how this section contributes to the goals, so removing this gives more time to focus on the kind of review system we're recommending - tool assisted review.

File: _episodes/41-code-review.md [hunk 3]
```diff
@@ -245,106 +176,8 @@ The following diagram depicts the branches that you should have in the repositor
...
-   based on their notification preferences.
-5. Once they accept the invitation, they will have the collaborator-level access to your repository
-   and will show up in the list of your collaborators.
-
-See the full details on
-[collaborator permissions for personal repositories](https://docs.github.com/en/account-and-profile/setting-up-and-managing-your-github-user-account/managing-user-account-settings/permission-levels-for-a-user-account-repository)
-to understand what collaborators will be able to do within your repository.
-Note that repositories owned by an organisation have a
-[more granular access control](https://docs.github.com/en/get-started/learning-about-github/access-permissions-on-github)
-compared to that of personal repositories.
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:272]: I restructured the examples so they don't require adding collaborators. I would be tempted to move this into the extras though, as it might be useful in some instances to know how to do this.

File: _episodes/41-code-review.md [hunk 4]
```diff
@@ -245,106 +176,8 @@ The following diagram depicts the branches that you should have in the repositor
...
-   Also note that, while we say "remote" branch `feature-x` -
-   you have actually obtained it locally on your machine when you cloned the remote repository.
-   ~~~
-   $ git checkout -b feature-x-tests origin/feature-x
-   ~~~
-   {: .language-bash}
-
-   You are now located in the new (local) `feature-x-tests` branch
-   and are ready to start adding your code.
-
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:310]: Now the exercises are reviewing a fixed changeset this is no longer required.

File: _episodes/41-code-review.md [hunk 5]
```diff
@@ -245,106 +176,8 @@ The following diagram depicts the branches that you should have in the repositor
...
-> checking if the requirements are satisfied.
-{: .callout}
-
-Remember to commit your new code to your branch `feature-x-tests`.
-
-~~~
-$ git add -A
-$ git commit -m "Added tests for feature-x."
-~~~
-{: .language-bash}
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:345]: As above - reader is no longer required to write some tests for someone else to review

File: _episodes/41-code-review.md [hunk 6]
```diff
@@ -373,17 +206,212 @@ you have to do the following:
...
+#### Bugs
+
+It is easier to make sure there are sufficient tests - you are not an accurate
+computer simulator anyway.
+
+#### Issues that pre-date the change
+
+You may spot something that the reviewer didn't introduce and think they could
+fix it while in the area, but this can be a rabbit hole. A better approach would be to
+raise a PR after this one has been merged fixing the thing you spotted.
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:381]: ❓ I'm not sure this is actually that useful so I'm tempted to just cut it.

File: _episodes/41-code-review.md [hunk 7]
```diff
@@ -465,6 +476,72 @@ and [Smartbear](https://smartbear.com/learn/code-review/best-practices-for-peer-
...
+* What kind of issues are (and aren't) appropriate to raise in a PR?
+* How will someone know when they are expected to take action (e.g. review a PR).
+
+Once you've introduced a review process, you should monitor (either formally or
+informally) how well it is working.
+
+It is important that reviews are processed quickly, to avoid costly context switching.
+We recommend aiming for 3 hours to get a first review, with the PR being merged the same
+day in most cases. If you are regularly missing these targets, then you should review
+where things are getting stuck and work out what you can do to move things along.
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:543]: Here I am assuming the reader is part of a team. However, I remember in discussions we talked how lots of users of this course are working independently. How do those scientists who are working independently use the code review section? Do they find someone else in their dept to review their code, reviewing it themselves, or just skipping this section?

**Review by thomaskileyukaea — COMMENTED** [review]
Explanation of why I removed certain sections / exercises as well as some ❓ questions for the maintainers that we should think about before merging this.

File: _episodes/41-code-review.md (hunk 6 above)
Comment (bielsnohr) [reply _episodes/41-code-review.md:381]: I sort of like this because it helps focus what code review should be about. That being said, it
isn't as important as the previous tips about what to look for, so perhaps we assess this based on
how long the section takes to deliver?

File: _episodes/41-code-review.md [hunk 8]
```diff
@@ -373,17 +206,212 @@ you have to do the following:
...
+> For each comment, either reply explaining why you don't think the change is necessary
+> or make the change and push a commit fixing it. You can reply to the comment indicating you
+> have done it.
+>
+> At the same time, people will be addressing your comments. If you're happy that your
+> comment has been suitably addressed, you can mark it as resolved. Once you're happy they
+> have all been addressed, you can approve the PR.
+{: .challenge}
 
 #### Step 6: Closing a Pull Request
```
Comment (bielsnohr) [inline _episodes/41-code-review.md:416]: Will need to get rid of "Step 6" and make sure the heading level is in line with the above.

File: _episodes/41-code-review.md [hunk 9]
```diff
@@ -465,6 +476,72 @@ and [Smartbear](https://smartbear.com/learn/code-review/best-practices-for-peer-
...
+is in a separate commit (i.e reformat the code, commit, refactor the code, commit, alter
+the behavior of the code, commit).
+
+Make sure you write a clear description of the content and purpose of the change.
+This should be provided as the pull request description.
+This should provide the context that reading the code will make more sense.
+
+It is also a good idea to review your code yourself. In doing this you will spot
+the more obvious issues with your code, allowing your reviewer to focus on the
+things you cannot spot.
```
Comment (bielsnohr) [inline _episodes/41-code-review.md:501]: The paragraphs need some semantic breaks at periods.

File: _episodes/41-code-review.md [hunk 10]
```diff
@@ -465,6 +476,72 @@ and [Smartbear](https://smartbear.com/learn/code-review/best-practices-for-peer-
...
+
+## Empathy in review comments
+
+Code is written by humans (mostly!), and code review is a form of communication. As such
+empathy is really important for effective reviewing.
+
+When reviewing code, it can be sometimes frustrating when code is confusing, particularly
+as it will be implemented differently to how you would have done it. However, it is important
+as a reviewer to be compassionate to the person whose code you are reviewing. Specifically:
```
Comment (bielsnohr) [inline _episodes/41-code-review.md:510]: Sembr

File: _episodes/41-code-review.md (hunk 7 above)
Comment (bielsnohr) [reply _episodes/41-code-review.md:543]: This is a good question. As someone who has lone coded in various scientific settings, I still use
merge/pull requests as a gate check on what I have just done. Of course, I haven't done that in
every instance, but it is a nice idea to strive for.

In terms of actual code review, this is tough. I was part of a community effort to try and define
what code review in research might look like: https://dev-review.readthedocs.io/en/latest/index.html

There, we did recommend trying to find someone to review your code. That person could be in your
research group (but just not working directly on what you are). We could link to that page here do
deal with the case of solo researchers working on code.

**Review by bielsnohr — COMMENTED** [review]
Overall, I really like where this is heading. The exercise based on a pre-determined changeset should help better achieve the objectives of this section. In the previous version, not having control over the incoming changes meant that learners might not see many of the points being raised in this section. So, this new approach gets my full endorsement.

It will be great to see how the section flows once the wording around this new exercise is added.

File: _episodes/41-code-review.md (hunk 2 above)
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:154]: I will make this adjustment as I think you're right, it is nice to have some other terms to look up, but with less detail.

File: _episodes/41-code-review.md [hunk 11]
```diff
@@ -245,106 +176,8 @@ The following diagram depicts the branches that you should have in the repositor
...
-$ git add -A
-$ git commit -m "Added tests for feature-x."
-~~~
-{: .language-bash}
-
-#### Step 4: Submitting a Pull Request
+### Raising a pull request
 
 When you have finished adding your tests
 and committed the changes to your local `feature-x-tests`,
```
[the remark this answers is not in the document]
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md]: Yes - this is now done.

File: _episodes/41-code-review.md [hunk 12]
```diff
@@ -213,281 +153,393 @@ you will need to specify the base branch and the head branch.
...
+
+1. Head over to the remote repository in GitHub
+   and locate your colleagues (`feature-std-dev`) branch from the dropdown box on the Code tab
+   (you can search for your branch or use the "View all branches" option).
+   ![All repository branches in GitHub](../fig/github-branches.png){: .image-with-shadow width="600px"}
+2. Open a pull request by clicking "Compare & pull request" button.
+   ![Submitting a pull request in GitHub](../fig/github-create-pull-request.png){: .image-with-shadow width="900px"}
+3. Select the base and the head branch, e.g. `main` and `feature-std-dev`, respectively.
+   Recall that the base branch is where you want your changes to be merged
+   and the head branch contains your changes.
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:186]: Make these instructions go to the pull requests page and specify the branches rather than relying on the time-sensitive button to appear.

File: _episodes/41-code-review.md [hunk 13]
```diff
@@ -213,281 +153,393 @@ you will need to specify the base branch and the head branch.
...
+4. Add a comment describing the nature of the changes,
+   and then submit the pull request.
+5. Repository moderator and other collaborators on the repository (code reviewers)
+   will be notified of your pull request by GitHub.
+6. At this point, the code review process is initiated.
+
+Find someone else to swap repositories with (i.e. you will review the pull request
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:193]: Maybe moving this content to after explaining what to look for would encourage the correct flow

File: _episodes/41-code-review.md [hunk 14]
```diff
@@ -213,281 +153,393 @@ you will need to specify the base branch and the head branch.
...
+
+Try to be as specific as you can, rather than "this code is unclear"
+prefer, "I don't understand what values this variable can hold".
+
+Make it clear in the comment if you want something to change as part
+of this PR. Ideally provide a concrete suggestion (e.g. better variable name).
+
+> ## Exercise: review some code
+>
+> Pick someone else in the group and go to the pull request they created on their repo.
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:262]: Explicitly show before hand how to add comments, and make explicit they should be added to the review but not submitted yet

File: _episodes/41-code-review.md [hunk 15]
```diff
@@ -213,281 +153,393 @@ you will need to specify the base branch and the head branch.
...
+#### Is the code a minimal change
+
+Does the code reimplement anything that already exists, either
+elsewhere in the codebase or in a library you know about?
+
+Does the code implement something that isn't on the ticket?
+
+#### Is the structure of the code clear
+
+Do functions do just one thing? Have appropriate design
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:237]: Could also talk about consistency of code here

File: _episodes/41-code-review.md [hunk 16]
```diff
@@ -213,281 +153,393 @@ you will need to specify the base branch and the head branch.
...
+### Raising a pull request for your fictional colleague
+
+1. Head over to the remote repository in GitHub
+   and locate your colleagues (`feature-std-dev`) branch from the dropdown box on the Code tab
+   (you can search for your branch or use the "View all branches" option).
+   ![All repository branches in GitHub](../fig/github-branches.png){: .image-with-shadow width="600px"}
+2. Open a pull request by clicking "Compare & pull request" button.
+   ![Submitting a pull request in GitHub](../fig/github-create-pull-request.png){: .image-with-shadow width="900px"}
+3. Select the base and the head branch, e.g. `main` and `feature-std-dev`, respectively.
+   Recall that the base branch is where you want your changes to be merged
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:185]: Here and elsewhere, github now refers to the head branch as the compare branch.

**Review by thomaskileyukaea — COMMENTED** [review]
Feedback from running through this with @bielsnohr Also noting we got to and mostly did responding to comments and it looks like it is 25min reading, 30min of exercises up to there. If the numbers at the top are accurate, it would seem this would take a little longer

File: _episodes/41-code-review.md [hunk 17]
```diff
@@ -213,281 +153,392 @@ you will need to specify the base branch and the head branch.
...
+* All paths through the code.
+* Making each `if` statement be evaluated as `True` and `False`.
+* Executing loops with empty, single and multi-element sequences.
+* Edge cases that you spot.
+* Any circumstances that you're not sure how certain code would behave.
 
-#### Step 5: Code Review
+Once you have built the list, go through the tests in the PR. Make sure
+the tests test what you expect (so inspect them closely!). Add a comment
+to the PR for any tests that are on your list that you can't find a suitable
```
Comment (thomaskileyukaea) [inline _episodes/41-code-review.md:314]: First use of PR should be explained

**Comment by thomaskileyukaea** [comment]
Finished off, looks like another 5min of reading, plus another 10 min of exercise.

File: _episodes/41-code-review.md [hunk 18]
```diff
@@ -208,286 +148,468 @@ and update your proposed changes within a self-contained bundle.
...
+Make sure your review comments are specific and actionable.
+
+Try to be as specific as you can, rather than "this code is unclear"
+prefer, "I don't understand what values this variable can hold".
+
+Make it clear in the comment if you want something to change as part
+of this pull request.
+Ideally provide a concrete suggestion (e.g. better variable name).
+
+## Adding a review comment
```
Comment (anenadic) [inline _episodes/41-code-review.md:256]: Feels like we need "what steps are of the code review?" to highlight the process earlier in the episode. So, e.g. reviewing the code and what to look for, commenting on code, buffering all comments and finishing/sending off our review, responding to review comments. Or something like that.

File: _episodes/41-code-review.md [hunk 19]
```diff
@@ -208,286 +148,468 @@ and update your proposed changes within a self-contained bundle.
...
+### Is there appropriate and up-to-date documentation
+
+If functionality has changed, has corresponding documentation been
+updated? If new functions have been added, do they have appropriate
+levels of documentation? Does the documentation make sense?
+
+Are there clear and useful comments that explain complex designs
+and focus on the "why/because" rather than the "what/how"?
+
+## Effective review comments
```
[the remark this answers is not in the document]
Comment (anenadic) [reply _episodes/41-code-review.md:245]: Also, "Effective review comments" could go under "adding a comment" either as the main text or as a callout.

File: _episodes/41-code-review.md [hunk 20]
```diff
@@ -208,286 +148,468 @@ and update your proposed changes within a self-contained bundle.
...
+>> for some data where the variance and standard deviation are different.
+>> In more complex examples, it is often easier to spot code that looks like it could  be wrong
+>> and think of a test that will exercise it. This saves embarrassment if the code turns out
+>> to be right, means you have the missing test written if it is wrong, and is often quicker
+>> than trying to execute the code in your head to find out if it is correct.
+>>
+> {: .solution}
+{: .challenge}
+
+## What not to look for
```
Comment (anenadic) [inline _episodes/41-code-review.md:423]: This section could be formatted as a bullet point list.

**Review by anenadic — COMMENTED** [review]
Thank you for the PR @thomaskileyukaea - apologies if took a while to review. We had a maintainers' meeting last week where we reviewed your changes. There is a couple of comments that can be addressed now (see my review of your PR) and then we'd like to have another meeting with you to discuss bigger changes which may be more difficult to explain in the PR's comments (and also because it affects some earlier episodes). If you are not available to meet and to work on this - we'd be happy to accept the PR to the dev branch and then apply our changes on top of yours, if that is OK with you.

File: _episodes/41-code-review.md [hunk 21]
```diff
@@ -65,32 +65,6 @@ and something you should adopt in your development process too.
 where one or several people from the team (different from the code's author)
 check the software by viewing parts of its source code.
 
-> ## Group Exercise: Advantages of Code Review
```
[the remark this answers is not in the document]
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:68]: Incorporated the question and some of the solution into the final exercise to reinforce the key messages of the lesson.

File: _episodes/41-code-review.md [hunk 22]
```diff
@@ -208,286 +148,468 @@ and update your proposed changes within a self-contained bundle.
...
-> Implement tests against the appropriate specification in your local feature branch.
+Your fictional colleague has implemented it according to the specification
+and pushed it to a branch `feature-std-dev`.
+You will turn this branch into a pull request for your fictional colleague on your
+repository.
+You will then engage in code review for the change (acting as a code reviewer) on
+a course mate's repository.
+Once complete, you will respond to the pull request on your repository from another team member.
+
+### Raising a pull request for your fictional colleague
```
[the remark this answers is not in the document]
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:176]: Unfortunately we believe the template approach is a bit of a non-starter for this due to the branch history being lost, meaning it isn't possible to create a PR from one branch to the other. Matt will discuss this with you in more detail at the next meeting to work out the best way forward. (I sadly won't be able to make it).

File: _episodes/41-code-review.md [hunk 23]
```diff
@@ -208,286 +148,468 @@ and update your proposed changes within a self-contained bundle.
...
+   ![Creating a new pull request.](../fig/github-create-pull-request.png){: .image-with-shadow width="900px"}
+6. Add a comment describing the nature of the changes,
+   and then submit the pull request by clicking `Create pull request`.
+![Submitting a pull request.](../fig/github-submit-pullrequest.png){: .image-with-shadow width="900px"}
+7. At this point, the code review process is initiated.
+
+We will now discuss what to look for in a code review,
+before practising this on this fictional change.
+
+## Things to look for in a code review
```
[the remark this answers is not in the document]
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:195]: I'm not sure I fully understand this comment, sorry. 

To add to the confusion, I recently changed how the collaborative code development models are structured on the main branch, this isn't merged into this branch(!). 

It is currently structured like:

> * Is the code readable
> * Is the code a minimal change
> ...
> 
> ### Is the code readable
> 
> More details about what this means
>
> ### Is the code a minimal change
>
> More details about what this means. 

Which, to my eyes, is the same as the collaborative code development models on the main branch. 

We _could_ structure like this:

> * Is the code readable - more details about what this means
> * Is the code a minimal change - more details about what this means
> ...

But to my eyes that is less clear as the bullet points become so long as to not really feel like bullet points. 

Alternatives I can see:

1. Make the bullet point list link to the sections so it is clear to the reader there is more detail to come
2. Make the bullet point list include a longer summary of each of the points, but still have the full sections after
3. Make the headings the bullet point list refers to smaller (e.g. just bold) so it looks more connected. 

Do let me know which of these you think would be best (or some combination of them). 

I'm not sure about the title change. I like that the use of the word "aim" emphasises that we're trying to get the code to a specific quality, but I worry it makes it ambiguous whether we're thinking about this from the reviewer or the authors perspective. How about:

* What should a code review aim to improve
* What are the qualities a code review should be aiming to improve

We do introduce the idea of design patterns in section 3, but I think modularity is a better thing to focus on anyway!

File: _episodes/41-code-review.md [hunk 24]
```diff
@@ -208,286 +148,468 @@ and update your proposed changes within a self-contained bundle.
...
+
+You can use `Add single comment` to immediately post a comment.
+However, it is best to batch the comments into a single review, so that the author
+knows when you have finished adding comments
+(and avoid spamming their email with notifications).
+
+Continue adding comments in this way, using the `Add review comment` button
+on subsequent comments.
+
+> ## Exercise: review some code
```
[the remark this answers is not in the document]
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:285]: I actually carefully setup this exercise to _not_ require collaborators (as you end up only making changes on your own repo, and making comments [which doesn't require collaborator permissions] on the other peoples). That said, it maybe is a bit complicated to see that - will add a diagram which will hopefully make it clearer.

File: _episodes/41-code-review.md (hunk 23 above)
[the remark this answers is not in the document]
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:195]: I have ended up doing at least (3) (though this will make 1 harder to do) as I think it helps with clarity.

File: _episodes/41-code-review.md (hunk 19 above)
[the remark this answers is not in the document]
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:245]: I've added

> Once a review has been raised it is over to the reviewer to review the code and submit a review.

At the start of the section on reviewing a pull request - is that what you meant?

File: _episodes/41-code-review.md (hunk 18 above)
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:256]: I've added a diagram in the code review in action section which will hopefully address this comment - let me know if you think we need more.

File: _episodes/41-code-review.md [hunk 25]
```diff
@@ -208,286 +148,468 @@ and update your proposed changes within a self-contained bundle.
...
+* All paths through the code.
+* Making each `if` statement be evaluated as `True` and `False`.
+* Executing loops with empty, single and multi-element sequences.
+* Edge cases that you spot.
+* Any circumstances that you're not sure how certain code would behave.
+
+Once you have built the list, go through the tests in the pull request. Make sure
+the tests test what you expect (so inspect them closely!).
+
+## Finishing a review
```
[the remark this answers is not in the document]
Comment (thomaskileyukaea) [reply _episodes/41-code-review.md:350]: I opted for putting the different review responses as a bullet point list, which I think makes sense, but maybe makes making the steps a list a bit much - let me know what you think

**Comment by thomaskileyukaea** [comment]
Thanks @anenadic for the review - it was really helpful. I've tried to address your comments - or else have replied with further clarifications. Do let me know if you have any further feedback.

I've merged the latest gh-pages into this so that it has the same collaborative code models.

**Comment by thomaskileyukaea** [comment]
I also looked into using Mermaid diagrams directly - at first it seemed tricky, but ended up being quite straight forward - here's a PR on my fork showing how it could be done: https://github.com/thomaskileyukaea/python-intermediate-development/pull/1

If we did go down this route, we could convert a few of the git diagrams into mermaid diagrams too, which would make maintenance easier.

**Comment by bielsnohr** [comment]
Comments from maintainer sprint:

- Concern about confusion when initially forking and having multiple branches. Solution: add some text around initial forking instructions about ignoring the additional branches
- The reasons for forking (TODO fill out) (i.e. need for retaining history in order to raise pull request)
- Options to retain templating:
  1. GitHub action that creates the required branch and changes upon creation of the copy from the template
  2. Have a script in the template repo that learners would run themselves (downside: system dependencies and learner error)
- obvious downsides are developer time to get these up and running, and then maintenance costs

**Comment by bielsnohr** [comment]
> I also looked into using Mermaid diagrams directly - at first it seemed tricky, but ended up being quite straight forward - here's a PR on my fork showing how it could be done: [thomaskileyukaea#1](https://github.com/thomaskileyukaea/python-intermediate-development/pull/1)
> 
> If we did go down this route, we could convert a few of the git diagrams into mermaid diagrams too, which would make maintenance easier.

Aha! Great that you were able to get this working. Of course, I should have remembered the solution of just injecting the required Javascript somewhere in the HTML. One gets quite used to everything being done with packages/gems/etc...

I would vote that we do go down the route of adding Mermaid. The diagram for this lesson looks great, see the bottom of this comment. GitHub already supports Mermaid, so the below is actually a rendering of the source code. Also, as @thomaskileyukaea mentions, we can convert other diagrams to Mermaid and it will make maintenance of the website easier (no more fiddling with svg, regenerating a png/jpeg, making sure they are in the right folder, etc).

Could I request that we get a maintainer vote on this? @anenadic I'll have a look through the governance document we are working on, but email list probably best?

```mermaid
%% Must disable useMaxWidth to have the diagram centre
%%{init: { 'sequence': {'useMaxWidth':false} } }%%
sequenceDiagram
    participant A as Author
    participant R as Reviewer
    A->>A: Write some code
    A->>R: Raise a pull request
    R->>R: Add comments to a review
    R->>A: Submit a review
    loop Until approved
        A->>R: Address or respond to review comments
        R->>A: Clarify or resolve comments
    end
    R->>A: Approve pull request
    A->>A: Merge pull request
```

**Comment by bielsnohr** [comment]
> we'd like to have another meeting with you to discuss bigger changes which may be more difficult to explain in the PR's comments (and also because it affects some earlier episodes). If you are not available to meet and to work on this - we'd be happy to accept the PR to the dev branch and then apply our changes on top of yours, if that is OK with you.

@anenadic I think because of various AL schedules, it has been difficult to arrange a synchronous call about the bigger changes you mention above. By "bigger changes" are you referring to the fact that steps in some of the initial episodes for obtaining the project would need to be modified?

If that is the extent of the changes, then I think it is quite manageable. There is the issue of regenerating some images of the GitHub interface, but I can assist with that.

As you suggest, it is probably best that we accept this MR in its current state into the `dev` branch and then continue working on the other changes earlier in the course ourselves (i.e. maintainers). I am keen to have Thomas get going on his proposed modifications to the Architecture and Design sections.

**Review by bielsnohr — APPROVED** [review]
As mentioned in comments above, I think we are fine to merge this into the dev branch and then take some time to make the edits to earlier parts of the lesson to advise on using forking.

**Comment by thomaskileyukaea** [comment]
✔️ Addressed issue where the schedule appeared at the bottom after the mermaid change.

**Comment by anenadic** [comment]
> Summary of outcome from Maintainer Meeting 2023-10-25:
> 
> * @anenadic and @steve-crouch have communicated that there is some content from the original version of the episode that they don't want to see removed, as has been done here
>   
>   * with @thomaskileyukaea having no more immediate time to work on this, the maintainers will take on the task of making any changes to get this updated episode to a state where it is ready to merge
> * it was decided that it probably isn't a good idea to merge this into `dev` since there are still things that need to be reviewed and changes to be made; therefore, this PR will stay open for the immediate future while this work is done
> * @bielsnohr will take on the task of modifying the earlier part of the lesson where a fork is now required, as stated in [Change initial retrieval of course project to be a fork rather than template copy #249](https://github.com/carpentries-incubator/python-intermediate-development/issues/249) .
>   
>   * this work will be pushed to a separate branch that stacks on top of this one

We can either re-create a dev branch for this PR or have a new branch of the same name `rework-review-section`. Will try to find some time to work on this with @steve-crouch - hopefully tomorrow. So we can have this merged as soon as possible. Apologies for not responding sooner - things have gone a bit hectic.

**Comment by bielsnohr** [comment]
> We can either re-create a dev branch for this PR or have a new branch of the same name `rework-review-section`. Will try to find some time to work on this with @steve-crouch - hopefully tomorrow. So we can have this merged as soon as possible. Apologies for not responding sooner - things have gone a bit hectic.

I think creating a new branch within our repo to hold this work is the best next step. I'll do that shortly. 

It would be good if we could collect the concrete items that need to change in this issue: #217
